docs/key-assignment/design.md

```diff
@@ -0,0 +1,149 @@
+# KeyAssignment v4
```

Comment:
Could we add a section in here that explains the two types of pruning we discussed? Would also be useful to define semantics besides "pruning" and "deletion"

Side note could explain that unbonded validator keys are not pruned unless they reassign

lmk if this already exists in the doc, I'm skimming during a zoom :)